File: src/lib/sync/protocol.test.ts
import { describe, expect, it } from 'vitest';

// Vite's `?raw`, the same way `format.test.ts` reads `shared/path-under.txt`:
// it keeps the fixture a build-time dependency rather than a path resolved at
// run time, so a file that moved fails the build instead of the assertion.
import FIXTURE_RAW from '../../../shared/sync-protocol.txt?raw';

import {
  CODE_LENGTH,
  PROTOCOL_FIELDS,
  emptyTimeline,
  formatCode,
  isErrorCode,
  normalizeCode,
  positionAt,
  shouldApply,
  type ContentRef,
  type Timeline,
} from './protocol';

// ---- the shared contract ----------------------------------------------------

/**
 * The half of the field-name contract this side owns.
 *
 * The compiler already proves `PROTOCOL_FIELDS` names every key of every
 * interface (see the `Complete<>` assertions in protocol.ts); this proves the
 * same lists match the ones Go produces. Both halves are needed — one without
 * the other is a contract with a single signature.
 */
describe('the wire matches shared/sync-protocol.txt', () => {
  const fixture = new Map<string, string[]>();
  for (const line of FIXTURE_RAW.split('\n')) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#')) continue;
    const [name, fields] = trimmed.split('\t');
    fixture.set(name.trim(), fields.trim().split(','));
  }

  // A file that moved, or was reformatted so every line reads as a comment,
  // would otherwise pass in silence — the one way a shared contract quietly
  // stops being one.
  it('parsed the fixture at all', () => {
    expect(fixture.size).toBeGreaterThanOrEqual(12);
  });

  it('names the same messages', () => {
    expect([...fixture.keys()].sort()).toEqual(Object.keys(PROTOCOL_FIELDS).sort());
  });

  for (const [name, want] of fixture) {
    it(`${name} has the agreed fields`, () => {
      expect([...(PROTOCOL_FIELDS[name] ?? [])].sort()).toEqual([...want].sort());
    });
  }
});

// ---- room codes -------------------------------------------------------------
//
// The same cases as `server/internal/wire/code_test.go`, because this is one
// behaviour written twice: the player normalises what somebody types so the
// relay is asked about a code that can exist, and the relay normalises again so
// a disagreement costs a refusal rather than the wrong room.

describe('normalizeCode', () => {
  it.each([
    ['ABC123', 'ABC123', 'already canonical'],
    ['abc123', 'ABC123', 'typed in lower case'],
    ['  ABC123  ', 'ABC123', 'pasted with space around it'],
    ['ABC-123', 'ABC123', 'read aloud, written with a dash'],
    ['A B C 1 2 3', 'ABC123', 'spelled out one character at a time'],
    // The whole reason for Crockford's alphabet: what a person types when they
    // are looking at the other glyph.
    ['ABCO23', 'ABC023', 'O read as zero'],
    ['ABCI23', 'ABC123', 'I read as one'],
    ['ABCl23', 'ABC123', 'lower-case L read as one'],
    ['', '', 'nothing'],
    ['ABC12', '', 'too short'],
    ['ABC1234', '', 'too long'],
    ['ABC!23', '', 'punctuation that is not a separator'],
    ['ABCU23', '', 'U — not in the alphabet and not a look-alike, so a typo'],
  ])('%s → %s (%s)', (input, want) => {
    expect(normalizeCode(input)).toBe(want);
  });

  it('is idempotent, or a code it produced would be refused on arrival', () => {
    for (const input of ['abc-123', 'ABCO23', 'ABCl23', ' zzz999 ']) {
      const once = normalizeCode(input);
      expect(once).not.toBe('');
      expect(normalizeCode(once)).toBe(once);
    }
  });

  it('formats for reading aloud, and leaves anything else alone', () => {
    expect(formatCode('ABC123')).toBe('ABC-123');
    expect(formatCode('short')).toBe('short');
    expect(normalizeCode(formatCode('ABC123'))).toBe('ABC123');
  });

  it('agrees with the length the UI promises', () => {
    expect(CODE_LENGTH).toBe(6);
  });
});

// ---- the projection ---------------------------------------------------------

describe('positionAt', () => {
  const content: ContentRef = { kind: 'url', url: 'x', title: 'x', duration: 100 };
  const playing: Timeline = {
    content,
    tracks: null,
    paused: false,
    position: 100,
    speed: 1,
    at: 10_000,
    rev: 1,
    by: 'a',
  };

  it('runs forward at the timeline speed', () => {
    expect(positionAt(playing, 12_000)).toBe(102);
    expect(positionAt({ ...playing, speed: 2 }, 12_000)).toBe(104);
    expect(positionAt({ ...playing, speed: 0.5 }, 12_000)).toBe(101);
  });

  it('does not advance while paused, however long ago the stamp was', () => {
    expect(positionAt({ ...playing, paused: true }, 999_000)).toBe(100);
  });

  it('does not advance when nothing is playing', () => {
    // `at` is stamped regardless, so without the content check the position of a
    // film that does not exist would run away.
    expect(positionAt({ ...emptyTimeline(), paused: false, at: 10_000 }, 999_000)).toBe(0);
  });

  it('runs backwards for a clock reading before the stamp — an offset is an estimate', () => {
    expect(positionAt(playing, 8_000)).toBe(98);
  });

  it('never comes out negative, because the caller seeks to it', () => {
    expect(positionAt(playing, -999_000)).toBe(0);
  });
});

describe('isErrorCode', () => {
  it('recognises what the relay can send', () => {
    expect(isErrorCode('no_room')).toBe(true);
    expect(isErrorCode('not_allowed')).toBe(true);
  });

  it('rejects anything else, so an unknown code cannot become a blank sentence', () => {
    expect(isErrorCode('teapot')).toBe(false);
    expect(isErrorCode('')).toBe(false);
  });
});

describe('shouldApply', () => {
  const at = (rev: number): Timeline => ({ ...emptyTimeline(), rev });

  it('takes a newer revision', () => {
    expect(shouldApply(at(3), at(4), false)).toBe(true);
  });

  it('refuses an older one', () => {
    // The failure this exists for is silent and permanent: a peer that applied a
    // stale snapshot sits at the wrong position with a healthy-looking room, and
    // nothing later contradicts it.
    expect(shouldApply(at(4), at(3), false)).toBe(false);
    expect(shouldApply(at(400), at(1), false)).toBe(false);
  });

  it('takes an equal revision, which is the correction after a refusal', () => {
    // `>=`, not `>`. A relay that refuses a guest's seek answers by re-sending
    // the room's current timeline at the revision it already had — and that is
    // exactly the message that pulls the guest back from the position they
    // optimistically moved to. Dropping it as "not newer" would leave them
    // somewhere the room is not, for good.
    expect(shouldApply(at(7), at(7), false)).toBe(true);
  });

  it('always takes the handshake', () => {
    // A `welcome` describes the room as it is now. It also covers the one case a
    // revision cannot: a room that ceased to exist and was created afresh counts
    // from zero again, and a client still holding the old count would otherwise
    // reject every timeline it was ever sent.
    expect(shouldApply(at(99), at(0), true)).toBe(true);
    expect(shouldApply(at(99), at(1), true)).toBe(true);
  });
});

File: server/internal/wire/protocol_test.go
package wire

import (
	"bufio"
	"encoding/json"
	"math"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"testing"
)

// The shared field-name contract. See shared/sync-protocol.txt for what this is
// defending against and why it is a file rather than a constant here.
func loadProtocolFixture(t *testing.T) map[string][]string {
	t.Helper()
	path := filepath.Join("..", "..", "..", "shared", "sync-protocol.txt")
	f, err := os.Open(path)
	if err != nil {
		t.Fatalf("open %s: %v", path, err)
	}
	defer f.Close()

	out := map[string][]string{}
	sc := bufio.NewScanner(f)
	for sc.Scan() {
		line := sc.Text()
		if strings.TrimSpace(line) == "" || strings.HasPrefix(strings.TrimSpace(line), "#") {
			continue
		}
		name, fields, ok := strings.Cut(line, "\t")
		if !ok {
			t.Fatalf("no tab in %q", line)
		}
		out[strings.TrimSpace(name)] = strings.Split(strings.TrimSpace(fields), ",")
	}
	if err := sc.Err(); err != nil {
		t.Fatal(err)
	}
	// A floor on the parse, for the same reason path-under.txt has one: a file
	// that moved, or was reformatted so every line reads as a comment, would
	// otherwise pass in silence — which is the one way a shared contract quietly
	// stops being one.
	if len(out) < 12 {
		t.Fatalf("parsed only %d messages from %s — did the file move or change shape?", len(out), path)
	}
	return out
}

// keysOf marshals a value and reports the field names it actually produced.
// That is the only reading of a Go struct that matches what the other side sees:
// a tag typo or a stray `omitempty` shows up here and nowhere in the type.
func keysOf(t *testing.T, v any) []string {
	t.Helper()
	raw, err := json.Marshal(v)
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}
	var m map[string]json.RawMessage
	if err := json.Unmarshal(raw, &m); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	out := make([]string, 0, len(m))
	for k := range m {
		out = append(out, k)
	}
	return out
}

func assertFields(t *testing.T, want []string, got []string, what string) {
	t.Helper()
	w := append([]string(nil), want...)
	g := append([]string(nil), got...)
	sort.Strings(w)
	sort.Strings(g)
	if strings.Join(w, ",") != strings.Join(g, ",") {
		t.Errorf("%s: fields are %v, shared/sync-protocol.txt says %v", what, g, w)
	}
}

func TestWireFieldsMatchSharedContract(t *testing.T) {
	// **This test must never be served from the build cache**, and getting that
	// wrong would quietly undo the whole contract.
	//
	// `go test` reuses a result when the package's own inputs are unchanged, and
	// the fixture lives outside this module — measured: rename a field in
	// `shared/sync-protocol.txt`, run `go test`, and it answers `(cached)` while
	// the two languages disagree. Which is precisely the failure this file
	// exists to prevent, wearing a different hat.
	//
	// `t.Setenv` marks a test uncacheable, which is the documented way to say
	// so. `-count=1` would work too and is what CI passes, but it is a flag a
	// person has to remember, and the whole point of a shared contract is that
	// it holds without anybody remembering anything.
	t.Setenv("FP_FIXTURE_IS_OUTSIDE_THIS_MODULE", "1")

	fx := loadProtocolFixture(t)
	yes := true

	tl := Timeline{Content: json.RawMessage(`{"kind":"url"}`), Paused: true, Position: 12.5, Speed: 1, At: 7, Rev: 3, By: "m1"}

	cases := []struct {
		name string
		v    any
	}{
		{"timeline", tl},
		{"member", Member{ID: "m1", Name: "n", Ready: true}},

		{"client:hello", ClientMsg{T: "hello", Ver: 1, Room: "ABC123", Name: "n"}},
		{"client:timeline", ClientMsg{T: "timeline", Timeline: &tl}},
		{"client:ready", ClientMsg{T: "ready", Ready: &yes, Reason: "buffering"}},
		{"client:mode", ClientMsg{T: "mode", HostOnly: &yes}},
		{"client:ping", ClientMsg{T: "ping", C: 99}},
		{"client:bye", ClientMsg{T: "bye"}},

		{"server:welcome", Welcome{T: "welcome", Ver: 1, Room: "ABC123", Me: "m1", Host: "m1",
			Members: []Member{}, Timeline: tl, Waiting: []string{}, Now: 1}},
		{"server:timeline", NewTimelineMsg(tl)},
		{"server:members", MembersMsg{T: "members", Members: []Member{}, Host: "m1", Waiting: []string{}}},
		{"server:pong", Pong{T: "pong", C: 1, S: 2}},
		{"server:error", ErrorMsg{T: "error", Code: "no_room", Message: "x"}},
	}

	for _, c := range cases {
		want, ok := fx[c.name]
		if !ok {
			t.Errorf("%s is not in shared/sync-protocol.txt", c.name)
			continue
		}
		assertFields(t, want, keysOf(t, c.v), c.name)
	}
	if len(cases) != len(fx) {
		t.Errorf("shared/sync-protocol.txt lists %d messages, this test covers %d", len(fx), len(cases))
	}
}

// ---- validation -------------------------------------------------------------

func TestValidateRejectsWhatARoomMustNeverSee(t *testing.T) {
	big := json.RawMessage("[" + strings.Repeat("0,", MaxContent/2) + "0]")
	tests := []struct {
		name string
		msg  ClientMsg
	}{
		{"unknown type", ClientMsg{T: "shrug"}},
		{"wrong version", ClientMsg{T: "hello", Ver: ProtocolVersion + 1}},
		{"code that is not a code", ClientMsg{T: "hello", Ver: 1, Room: "no"}},
		{"timeline with no timeline", ClientMsg{T: "timeline"}},
		{"ready with no flag", ClientMsg{T: "ready"}},
		{"mode with no flag", ClientMsg{T: "mode"}},
		{"NaN position", ClientMsg{T: "timeline", Timeline: &Timeline{Position: math.NaN(), Speed: 1}}},
		{"infinite position", ClientMsg{T: "timeline", Timeline: &Timeline{Position: math.Inf(1), Speed: 1}}},
		{"negative position", ClientMsg{T: "timeline", Timeline: &Timeline{Position: -1, Speed: 1}}},
		{"absurd position", ClientMsg{T: "timeline", Timeline: &Timeline{Position: MaxPosition + 1, Speed: 1}}},
		{"zero speed", ClientMsg{T: "timeline", Timeline: &Timeline{Speed: 0}}},
		{"speed past the ceiling", ClientMsg{T: "timeline", Timeline: &Timeline{Speed: MaxSpeed + 1}}},
		{"NaN speed", ClientMsg{T: "timeline", Timeline: &Timeline{Speed: math.NaN()}}},
		{"oversized content", ClientMsg{T: "timeline", Timeline: &Timeline{Content: big, Speed: 1}}},
	}
	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			msg := tc.msg
			if err := msg.Validate(); err == nil {
				t.Fatalf("accepted %+v", tc.msg)
			}
		})
	}
}

func TestValidateNormalises(t *testing.T) {
	msg := ClientMsg{T: "hello", Ver: 1, Room: " abc-123 ", Name: "  a\x00b  "}
	if err := msg.Validate(); err != nil {
		t.Fatalf("rejected: %v", err)
	}
	if msg.Room != "ABC123" {
		t.Errorf("room = %q, want ABC123", msg.Room)
	}
	if msg.Name != "ab" {
		t.Errorf("name = %q, want %q — control characters must not reach other viewers", msg.Name, "ab")
	}
}

// A JSON `null` and an absent field both mean "nothing is playing", and every
// later `Content == nil` depends on the two having been collapsed into one.
func TestExplicitNullContentBecomesNil(t *testing.T) {
	var msg ClientMsg
	if err := json.Unmarshal([]byte(`{"t":"timeline","timeline":{"content":null,"position":0,"speed":1}}`), &msg); err != nil {
		t.Fatal(err)
	}
	if err := msg.Validate(); err != nil {
		t.Fatal(err)
	}
	if msg.Timeline.Content != nil {
		t.Errorf("content = %q, want nil", msg.Timeline.Content)
	}
}

// ---- projection -------------------------------------------------------------

func TestPositionAt(t *testing.T) {
	content := json.RawMessage(`{"kind":"url"}`)
	playing := Timeline{Content: content, Position: 100, Speed: 1, At: 10_000}

	if got := playing.PositionAt(12_000); got != 102 {
		t.Errorf("two seconds later = %v, want 102", got)
	}
	if got := (Timeline{Content: content, Position: 100, Speed: 2, At: 10_000}).PositionAt(12_000); got != 104 {
		t.Errorf("at double speed = %v, want 104", got)
	}
	paused := playing
	paused.Paused = true
	if got := paused.PositionAt(999_000); got != 100 {
		t.Errorf("paused = %v, want 100 — a paused timeline must not advance", got)
	}
	// Nothing playing: `at` is still stamped, so without the content check the
	// position would run away from a film that does not exist.
	empty := Timeline{Position: 0, Speed: 1, At: 10_000}
	if got := empty.PositionAt(999_000); got != 0 {
		t.Errorf("empty = %v, want 0", got)
	}
	// A clock reading earlier than the stamp is legitimate — a client's offset
	// is an estimate — and the projection simply runs backwards with it.
	if got := playing.PositionAt(8_000); got != 98 {
		t.Errorf("two seconds before the stamp = %v, want 98", got)
	}
	// But it must never come out negative, because the caller seeks to it.
	if got := playing.PositionAt(-999_000); got != 0 {
		t.Errorf("far behind the stamp = %v, want 0", got)
	}
}
